lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,1004 @@
+    }
+  }
+  // Before connect nodes, need to mark all back edges first
+  detectBackEdges(FRG, F);
+  // Connect nodes between different basic blocks
+  for (auto &BB : *F) {
+    auto *BBI = FRG->getBasicBlockHelperInfo(&BB);
+    auto *Term = BB.getTerminator();
+    // Check all the connections from BB in CFG
+    for (const auto *SB : Term->successors()) {
+      auto *SBI = FRG->getBasicBlockHelperInfo(SB);
+      ExecutionCountType C = 0;
+      auto BBCount = StructInfo->getExecutionCount(&BB);
+      auto SBCount = StructInfo->getExecutionCount(SB);
+      if (BBCount.hasValue() && SBCount.hasValue()) {
+        // Take probability of the branch
+        auto Prob = StructManager->getBranchProbability(&BB, SB);
+        assert(Prob.hasValue());
+        C = std::min(BBCount.getValue() * Prob.getValue(),
+                     SBCount.getValue() * 1.0);
+      }
+      if (!DisableIgnoreZeroCountNodes && C < 1e-3)
+        continue;
+      auto D = BBI->RemainBytes; // Use size of remaining data in BB
+      auto isBackEdge = (BBI->BackEdgeSet.find(SB) != BBI->BackEdgeSet.end());
+      FRG->connectNodes(BBI->LastNode, SBI->FirstNode, C, D, isBackEdge);
+    }
+  }
+  assert(StructInfo->getExecutionCount(&F->getEntryBlock()));
+  auto *BBI = FRG->getBasicBlockHelperInfo(&F->getEntryBlock());
+  FRG->setRootNode(BBI->FirstNode);
+  DEBUG(dbgs() << "---------- FRG for function " << F->getName()
+               << "----------------\n");
+  DEBUG(FRG->debugPrint(dbgs()));
+  DEBUG(dbgs() << "------------------------------------------------------------"
+                  "------------\n");
+  return FRG;
+}
+
+// Core function to update a cell in CPT.
+// In this function, a CPT can be either CloseProximityTable (real CP relations)
+// or GoldCPT (used for result validation). A cell in CPT represents CP
+// relations between two fields and consists of a pair of numbers: count and
+// distance. When updating a cell, call updatePairByMerging().
+void CloseProximityBuilder::updateCPT(FieldNumType Src, FieldNumType Dest,
+                                      ExecutionCountType C, DataBytesType D,
+                                      CloseProximityTableType &CPT) {
+  assert(Src <= NumElements && Dest <= NumElements);
+  if (Src == 0 || Dest == 0 || C < 1e-3 || Src == Dest)
+    // Give up update if from or to a dummy node, or if the count is zero
+    return;
+  assert(!std::isnan(C));
+  if (Src != 0 && Dest != 0) {
+    if (Src > Dest) {
+      std::swap(Src, Dest);
+    }
+
+    DEBUG_WITH_TYPE(
+        DEBUG_TYPE_CPG,
+        dbgs() << " Before update: "
+               << "(" << DEBUG_PRINT_COUNT(CPT[Src - 1][Dest - 1].first) << ","
+               << DEBUG_PRINT_DIST(CPT[Src - 1][Dest - 1].second) << ")\n");
+    assert(CPT[Src - 1][Dest - 1].first + C != 0);
+    updatePairByMerging(CPT[Src - 1][Dest - 1].first,
+                        CPT[Src - 1][Dest - 1].second, C, D);
+    if (std::isnan(CPT[Src - 1][Dest - 1].first) ||
```

Comment:
Should've removed it. It was many versions ago when there's a possibility that 0/0 happens. I'll remove it now...